### ISSUE TITLE
`import/no-extraneous-dependencies` and `import/order` false postives since version 2.23

### ISSUE THREAD

**Issue by mlbiche** [benmosher/eslint-plugin-import#2065]
Since the update to version 2.23, we have a lot of false positives in our project being part of two categories :
 
* `error  'rxjs/operators' should be listed in the project's dependencies. Run 'npm i -S rxjs/operators' to add it  import/no-extraneous-dependencies` : `rxjs/operators` is a sub-module of `rxjs` so I don't have to add it to project dependencies.
*```error  `@models/partner/partner-id` import should occur before import of `@models/partner/partner`               import/order``` : I can fix these issues, but it seems that the order has changed between two versions and I don not understand why and if it is on purpose ? I can not find mention to this in the changelogs.
* 
Cheers 🤙

**Comment by tarlepp** [comment]
I'm having this same issue with `rxjs/operators` any solution to fix that?

**Comment by ljharb** [comment]
cc @paztis, #1696 may have caused this first issue.

**Comment by ljharb** [comment]
@mlbiche for the `order` issue, can you provide the exact code it's warning on?

**Comment by mlbiche** [comment]
@ljharb Here it is :

```
import { Partner } from '@models/partner/partner';
import { PartnerId } from '@models/partner/partner-id';
```

It is doing the same whenever there are two files at the same path with a common prefix. Before, the shortest one was considered first.

**Comment by tarlepp** [comment]
Updated to `"eslint-plugin-ngrx": "1.23.0"`  and I 'm still getting;
```
/app/src/app/app.component.ts
  7:1  error  'rxjs/operators' should be listed in the project's dependencies. Run 'npm i -S rxjs/operators' to add it  import/no-extraneous-dependencies
```

**Comment by ljharb** [comment]
#2071 should fix the ordering issue.

**Comment by lensbart** [comment]
> #2071 should fix the ordering issue.

Not sure if covered by this PR, but the following comes up as an error since upgrading to 2.23:

```ts
import { UserInputError } from 'apollo-server-express'

import { new as assertNewEmail } from 'Assertions/Email'
```

```json
// .eslintrc.json (abridged)
{
  "plugins": [
    "import",
  ],
  "rules": {
    "import/order": [
      "error",
      {
        "alphabetize": {
          "caseInsensitive": true,
          "order": "asc"
        },
        "groups": [
          "builtin",
          "external",
          "internal",
          "parent",
          "sibling",
          "index"
        ],
        "newlines-between": "always"
      }
    ],
  },
  "settings": {
    "import/resolver": {
      "babel-module": {},
      "webpack": {}
    }
  }
}

```

The codebase I’m working on uses the convention that aliases start with a capital letter. There should be a space between import groups. The error goes away when removing the empty line between both import statements.

**Comment by ljharb** [comment]
@lensbart Could you file that as a separate issue?

**Comment by lensbart** [comment]
Sure thing! https://github.com/benmosher/eslint-plugin-import/issues/2081

**Comment by paztis** [comment]
Is the defect still present ?
EWHat is the resolver plugin you use ? I don't encounter the issue on my side

**Comment by paztis** [comment]
ah ok I see the problem.
With the new mechanism, we search for hte fist packageJson ancestor

the problem is rxjs /operators/package.json exists with this content
```
{
  "name": "rxjs/operators",
  "typings": "./index.d.ts",
  "main": "./index.js",
  "module": "../_esm5/operators/index.js",
  "es2015": "../_esm2015/operators/index.js",
  "sideEffects": false
}
```

So it isconsidered as an independent module, what is not the case...
It is an error of the rxjs developers I think

What can we do for this ?

**Comment by clydin** [comment]
That is the `package.json` for a secondary entry point of the package. Until the `exports` field becomes more widely supported these files are unfortunately going to be around.

To find the root `package.json` for a particular package, the `node_modules` directory could to be used in the check.
For example:
* `node_modules/<unscoped_package>/package.json`
* `node_modules/<@scope>/<scoped_package>/package.json`

Alternatively, the `package.json` could be read and validated for all required fields but that would probably be more expensive.  There is no `version` in the above package.json, for example.

**Comment by ljharb** [comment]
@paztis package.json files are only the root of a project when `../` is named "node_modules".

**Comment by paztis** [comment]
No, this is not true. I do this recursive search explicitely to support the. Monolithic repositories (multiple packages in one repository).
If we do like you said, we will always found a parent package.json  my hat is totally unrelated with the package.

Standards are 1 package.json for each delivreable

**Comment by paztis** [comment]
In case we of monorepo, what is possible is to resolve the import file without resolving the symbolic links. Potentially this may also us to check the node_module path. We still have the case of scoped modules to verify

**Comment by tarlepp** [comment]
> ```
> {
>   "name": "rxjs/operators",
>   "typings": "./index.d.ts",
>   "main": "./index.js",
>   "module": "../_esm5/operators/index.js",
>   "es2015": "../_esm2015/operators/index.js",
>   "sideEffects": false
> }
> ```

Note that this in `6.x` branch and not in main (7.x) anymore - not sure if this error occurs with `7.x` version - cannot test that because I'm using RxJs with Angular and it doesn't support RxJs 7.x at this time.

**Comment by paztis** [comment]
solution I propose in #2072 might resolve your problem. It's will be a little less strict on the imports but might be enough for these cases

**Comment by ljharb** [comment]
@paztis i said an _installed_ project. In other words, if you're inside a node_modules directory, then you're not in a monorepo, you're an npm-installed package, and the rule I described applies.

**Comment by paztis** [comment]
I  really understand your problem.
But what you misunderstood is that in a mini repo, under node_modules you have all the modules of you monorepo in symlink.

So when you resolve the filename of the imports, the resolver can give you the real path of the file that is not under node_modules.
This is the principle of monorepos

And as I can it control the resolvers as they are plugins of eslit-import (lot of plugins exist on nom), I won't just check if I'm under node module or not

**Comment by ljharb** [comment]
@paztis yes, but the symlink will be realpathed, so the resulting paths won't be under node_modules.

**Comment by ljharb** [comment]
Either way, "supporting a monorepo" is of zero importance compared to "supporting normally installed dependencies", if there's a conflict between the two.

**Comment by paztis** [comment]
Sorry but the only non standard thing here is the rocks module that produces abnormal package.json in its subfolders.

Now I've propose you a solution that will resolve your errors. So do you want it or not ?

**Comment by ljharb** [comment]
Nested package.jsons aren't abnormal, they are an explicitly encouraged pattern with node's native ESM support.

I'll take a look at your comment in #2072.

**Comment by paztis** [comment]
To detect the module package from one package file, we're using read-pkg-up, downloaded 38 228 565 times on npm.
It recursively move up until it founds a package.json
So it means everyone that use this module is not correctly discovering the packages ?

**Comment by paztis** [comment]
Esm in node is really recent. Monorepo are older.
Need to found one solution to make both cohabit

**Comment by ljharb** [comment]
Yes, if they're using read-pkg-up and stopping at the first package.json AND using it as the project root, then they all have this bug.

Nested package.jsons aren't recent; node has always supported nested package.json files with a "main". The issue is that the closest package.json hasn't ever been guaranteed to be the package root, so that's the bug that needs fixing.

**Comment by piranna** [comment]
> Yes, if they're using read-pkg-up and stopping at the first package.json AND using it as the project root, then they all have this bug.

From its [readme](https://www.npmjs.com/package/read-pkg-up):

> Read the closest package.json file

To me, that's not a bug, but instead the expected behaviour.

**Comment by ljharb** [comment]
Of that package, absolutely. However, anyone _naively_ using it to find the root package.json for a package is what has a bug.

**Comment by piranna** [comment]
But, how to detect the actual package root?

**Comment by ljharb** [comment]
@piranna there isn't a robust way for things not inside a node_modules folder. There's reliable *signals* - if it has "exports", or "publishConfig", or "private", I'd call it top-level. If the repo has a ".git" directory, it's probably top-level. It'd be a heuristics-based approach.

**Comment by JFGHT** [comment]
I am having this very same problem with Firebase imports.

**Comment by paztis** [comment]
I've proposed a fix for extraneous-dependencies in PR #2097

**Comment by Temez1** [comment]
Still happens with firebase
![image](https://user-images.githubusercontent.com/46352445/135605489-5ca3a5b9-b74e-4685-8582-8bdf62ee5b08.png)

`"eslint-plugin-import": "^2.24.2",`

**Comment by latipun7** [comment]
> Still happens with firebase ![image](https://user-images.githubusercontent.com/46352445/135605489-5ca3a5b9-b74e-4685-8582-8bdf62ee5b08.png)
> 
> `"eslint-plugin-import": "^2.24.2",`

did you list `@firebase/auth` in your project `dependencies` or `devDependencies`? If not, then it's expected result. Even if you have `@firebase/auth` in your `node_modules` as part of the another module's dependencies.

This issue happened mainly for sub module or sub files of module detected as individual module. While [`@firebase/auth`](https://www.npmjs.com/package/@firebase/auth) is individual module.
For example [`rxjs`](https://www.npmjs.com/package/rxjs) is individual module that exported `operators.js` in that module. So we can import `rxjs/operators.js` or `rxjs/operators` which is valid as `rxjs` as dependencies.

**Comment by Temez1** [comment]
Oh, I didn't know about the scope thing at all. And I don't know why firebase is using separate modules as its own packages when it says it's not supposed to be used directly :D 

Anyways, all good when I removed the "@" and used `firebase/auth` instead.

Thanks for the info.
